Live preview

Nitwel opened the pull request:
This feature adds live previewing capabilities of your website to Directus.

https://user-images.githubusercontent.com/22577866/231797682-3d877ec4-a2a2-44a5-a3cb-156753a572d2.mp4

Nitwel committed:
Add split view to the private view

Nitwel committed:
add more functionality to split view

Nitwel committed:
change text

Nitwel committed:
add simple size settings

Nitwel committed:
finish basic live preview feature

Nitwel committed:
add reload functionality

br41nslug commented:
Im thinking that perhaps we should automatically add the domain used in the preview config to the CSP whitelist. Because manually updating the `CONTENT_SECURITY_POLICY_DIRECTIVES__FRAME_SRC=http://my-cool-domain.net` environment variable then restarting Directus for it to take effect does not seem like the best user experience to me. And is not possible on cloud at all at this time.

paescuj commented:
> Im thinking that perhaps we should automatically add the domain used in the preview config to the CSP whitelist. Because manually updating the `CONTENT_SECURITY_POLICY_DIRECTIVES__FRAME_SRC=http://my-cool-domain.net` environment variable then restarting Directus for it to take effect does not seem like the best user experience to me. And is not possible on cloud at all at this time.

Yes, that would be exactly the idea 👍

Nitwel committed:
update pnpm-lock

Nitwel committed:
Merge remote-tracking branch 'origin/main' into live-preview

Nitwel committed:
update to newer design

Nitwel committed:
remove console.log

Nitwel committed:
Fix height of main content

Nitwel committed:
update design

Nitwel committed:
finishing fixes

Nitwel committed:
run linter

Nitwel committed:
Merge remote-tracking branch 'origin/main' into live-preview

Nitwel committed:
finishing cleanup

ConnorSimply commented:
> Will this feature open the website preview with some kind of `token` like `https://example.com?token=sljdfksdjflsdf`
> 
> To give the ability to fetch data that is not publicly available yet? Because this feature makes most sense to show changes that are not yet published right?

This currently functions by adding query params to the page url via a template you set and control in the collection settings.
<img width="802" alt="image" src="https://user-images.githubusercontent.com/12714889/232384329-40835f53-2961-455d-a1fc-d56f327ffab2.png">

azrikahar committed:
Merge branch 'main' into live-preview

azrikahar committed:
rebuild lock file

azrikahar committed:
lint fix

rijkvanzanten committed:
Merge branch 'main' into live-preview

br41nslug reviewed (commented):
Im thinking the right side of the frame should have a border too (or the border should toggle dark mode with the rest of the interface)
![image](https://user-images.githubusercontent.com/9389634/232552924-926546cd-a91c-4666-87cc-81e9753dd821.png)

The scale size is not selectable when not in mobile mode:
![image](https://user-images.githubusercontent.com/9389634/232553258-5a620ebd-6428-40d3-b7e9-35c54dafaabf.png)

The refresh button is missing a loading interaction (or any indication) to show that anything has actually happened:

https://user-images.githubusercontent.com/9389634/232553547-8346e310-9eb5-43fb-a16f-cb4f340f1612.mp4

LasseRosenow commented:
> This currently functions by adding query params to the page url via a template you set and control in the collection settings.

So for now this would only allow me to maybe show items that are drafts still, but no real way to preview changes of public items before they are applied?

I guess this would maybe be added once #2975 is implemented?

I was hoping maybe for now in the live preview the current state could be saved in a JSON field in the db for example and be send to the client instead of the real values, if for example ?preview=true is applied to the http request 🤔

This way it would be really convenient to see changes in the live preview before actually applying them

azrikahar committed:
hide preview for collections without preview urls

azrikahar committed:
fix refresh and eslint error

rijkvanzanten commented:
> I guess this would maybe be added once https://github.com/directus/directus/discussions/2975 is implemented?

This is the plan as of right now 🙂 The two features really enhance eachother, but it makes more sense to implement them as two different sprints 👍🏻 (otherwise this'll turn into another 6-month PR 😅 )

azrikahar committed:
use v-text-overflow for preview url

azrikahar committed:
use bottom placement for consistency

because other buttons such as Open in New Window and Refresh shows their tooltip at the bottom.

azrikahar committed:
tiny code tweak with early return

Nitwel committed:
remove fun parts and change divider color

azrikahar committed:
Merge branch 'main' into live-preview

azrikahar committed:
fix preview for singletons

azrikahar committed:
Merge branch 'main' into live-preview

azrikahar committed:
slightly improve the refresh button feedback

azrikahar committed:
hide zoom in fullscreen/desktop view

it is disabled in fullcreen, but some users may think they can click it, so I've opted to hide it instead. Especially since switching from mobile to desktop mode always resets it to 100%.

azrikahar committed:
tiny early return tweak

azrikahar committed:
opt to fix v-select disabled state cursor

azrikahar committed:
reverse zoom scaling order

azrikahar committed:
prevent preview going out of bounds when zoomed

azrikahar committed:
fix dragging when pointer enters iframe

azrikahar committed:
reinstate previous code tweak

azrikahar committed:
Merge branch 'main' into live-preview

azrikahar committed:
regenerate lock file

azrikahar committed:
use lockfile from main

paescuj commented:
> > Im thinking that perhaps we should automatically add the domain used in the preview config to the CSP whitelist. Because manually updating the `CONTENT_SECURITY_POLICY_DIRECTIVES__FRAME_SRC=http://my-cool-domain.net` environment variable then restarting Directus for it to take effect does not seem like the best user experience to me. And is not possible on cloud at all at this time.
> 
> Yes, that would be exactly the idea 👍

@Nitwel Can you take a look at this? 🙏

Nitwel committed:
Merge remote-tracking branch 'origin/main' into live-preview

paescuj committed:
Dear linter...

paescuj committed:
Merge branch 'main' into live-preview

LasseRosenow commented:
![Screenshot_20230509-164658_Chrome.png](https://github.com/directus/directus/assets/10547444/e990e8f8-10dd-4949-903b-292a600c003a)

I was wondering, is this the final design of the user interface?

Because for me at least the white bar with the black buttons inside feel a bit alien compared to the rest of the application.

Also the white border between the live preview and the webapp feels too bright, maybe replace it with a primary color version?

paescuj committed:
Merge branch 'main' into live-preview

estheragbaje committed:
chore: add live preview to sidebar

estheragbaje committed:
docs: add frontmatter

estheragbaje committed:
docs: add intro

estheragbaje committed:
docs: before you start

estheragbaje committed:
docs: live preview in directus

estheragbaje committed:
docs: draft mode in nextjs

estheragbaje committed:
docs: fetch with draft mode

estheragbaje committed:
docs: preview content in directus

estheragbaje committed:
docs: next steps

estheragbaje committed:
docs: update

estheragbaje committed:
docs: updates

estheragbaje committed:
docs: add live preview to collections

estheragbaje committed:
docs: add live preview to item

estheragbaje committed:
docs: updates

estheragbaje committed:
docs: update

phazonoverload commented on docs/app/content/items.md:
```diff
@@ -192,3 +192,28 @@ If you wish to remove an Item from view or use without fully deleting it, consid
+5. Toggle **"Enable Preview"** at the top of the page.
+6. Make changes to your Item and save. The preview will appear on the right side of your screen.
+
+:::tip Preview URL Format
+
+The format of the preview URL in Directus is: `http://your-directus-url/collection/ID`
+
+:::
```
What do you mean by this?

phazonoverload commented on docs/app/data-model/collections.md:
```diff
@@ -380,3 +379,23 @@ To delete a collection, follow these steps.
 
 :::
+
+## Live Preview a Collection
```
```suggestion
## Set Up Live Preview For a Collection
```

phazonoverload commented on docs/app/data-model/collections.md:
```diff
@@ -380,3 +379,23 @@ To delete a collection, follow these steps.
+To enable this feature, navigate to the **Settings** menu and select **Data Model.** Choose the collection you want to
+allow the live preview feature.
+
+Add your URL as the **Preview URL** in the collection settings and select the `ID`. For example, the preview URL of a
+posts collection running on localhost:3000 will be `http://localhost:3000/posts/ID`.
```
Again I think this is misleading. A user can set up any url pattern they want with dynamic values from the specific item they are previewing. I think this should be mentioned here.

phazonoverload commented on docs/app/data-model/collections.md:
```diff
@@ -380,3 +379,23 @@ To delete a collection, follow these steps.
+Add your URL as the **Preview URL** in the collection settings and select the `ID`. For example, the preview URL of a
+posts collection running on localhost:3000 will be `http://localhost:3000/posts/ID`.
+
+Now, click **"Enable Preview**" option at the top of the page. Whenever you make changes to the data in your
+collection and save them, you should see a live preview of the item on the right-hand side of the screen.
+
+:::tip Preview on Different Screen Sizes
+
+Toggle the **"Dimensions Display"** icon to preview your content on desktop or mobile screens.
+
+:::
```
This should go on the items page.

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+
+## Configure Live Preview URL in Directus
+
+Go into your Directus instance, create a new collection, and give it a name such as "Posts". Add the necessary fields to
+your collection, such as "Title" and "Content."
```
```suggestion
In your Directus instance, create a new `Posts` collection. Add `title` and `content` fields to your collection.
```

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+Go into your Directus instance, create a new collection, and give it a name such as "Posts". Add the necessary fields to
+your collection, such as "Title" and "Content."
+
+> Skip to the next step if you already have a collection with the necessary fields.
```
Remove this. I do see what you're doing, but it is ok to write opinionated guides for the docs to help people get the answer they need. I think folks will know they can sub out this collection with any other.

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+specify the Preview URL for your Next.js project by selecting ID from the dropdown and entering a URL in this format:
+`http:/<your-site>/api/draft?secret=MY_SECRET_TOKEN&id=ID`
+
+Make sure to replace “MY_SECRET_TOKEN’ with the secret you want in your Next.js project and save your changes.
```
```suggestion
Make sure to replace `MY_SECRET_TOKEN` with the secret you want in your Next.js project and save your changes.
```

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+
+Navigate to Settings -> Data Model and select the collection you want to configure. In the "Preview URL" section,
+specify the Preview URL for your Next.js project by selecting ID from the dropdown and entering a URL in this format:
+`http:/<your-site>/api/draft?secret=MY_SECRET_TOKEN&id=ID`
```
Question - do they need to ensure `ID` is inserted as a variable in the UI? This might be a good point for screenshots.

estheragbaje replied:
Yes, they need to. Exactly, this is one of the places I'll add a GIF to

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+collection or adding new content require rebuilding the entire site for the changes to take effect. With Draft Mode
+enabled, pages can be rendered at request time instead of build time.
+
+In your Next.js application, create a route handler such as `app/api/draft/route.ts` and include the following code:
```
```suggestion
In your Next.js application, create a route handler file at `app/api/draft/route.ts` and include the following code:
```

phazonoverload replied:
"such as" can be removed from guides generally as we are aiming for an authoritative tone.

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
````diff
@@ -0,0 +1,155 @@
+}
+```
+
+This code sets the `secret` variable to "MY_SECRET_TOKEN" and validates whether the `secret` parameter in the request
````
```suggestion
This code sets the `secret` variable to `MY_SECRET_TOKEN` and validates whether the `secret` parameter in the request
```

phazonoverload commented on docs/guides/headless-cms/live-preview/nextjs.md:
```diff
@@ -0,0 +1,155 @@
+
+## Preview Content in Directus
+
+Go over to an item in your collection and toggle "Enable Preview" at the top of the page. Whenever you create or edit an
+item in your collection and “click” save, you should see a live preview of the item on the right-hand side of the
+screen.
```
```suggestion
In an item page, toggle "Enable Preview" at the top of the page. Whenever you create or edit an
item in your collection and “click” save, you should see a live preview of the item on the right-hand side of the
screen.
```

phazonoverload reviewed (commented):
@estheragbaje your Next guide looks great. A couple of minor changes on that and the location of information on the app guide related to this feature. Overall, fantastic.

phazonoverload reviewed (commented):
@estheragbaje your Next guide looks great. A couple of minor changes on that and the location of information on the app guide related to this feature. Overall, fantastic.

estheragbaje committed:
chore: updates to the nextjs guide

estheragbaje committed:
chore: updates to collections

estheragbaje committed:
chore: updates to items

phazonoverload committed:
Nuxt 3 guide

phazonoverload committed:
Small Next guide fixes

phazonoverload committed:
Added tips

paescuj committed:
Merge branch 'main' into live-preview

LasseRosenow commented:
One thing I am wondering about:
Is it possible to specify a base-url for the preview_urls?

So I don't have to specify my whole website domain for every collections preview URL?
This can especially be time consuming and dangerous to update when migrating from the development to the production deployment.

paescuj committed:
Spellcheck fixes

phazonoverload commented:
> One thing I am wondering about: Is it possible to specify a base-url for the preview_urls?

Not in this implementation

azrikahar committed:
Merge branch 'main' into live-preview

paescuj committed:
Rework resizement

br41nslug commented on app/src/components/v-resizeable.vue:
```diff
@@ -0,0 +1,216 @@
+</template>
+
+<script setup lang="ts">
+// TODO
+// - Hide "grab-bar" if wrapper itself is not in viewport
+//   Ref: https://github.com/directus/directus/pull/12050
+// - Move debounce to parent
```
Is this still TODO?

paescuj replied:
Yes 👍

br41nslug commented on app/src/modules/content/routes/item.vue:
```diff
@@ -57,6 +57,18 @@
 		</template>
 
 		<template #actions>
+			<v-button
+				v-if="previewURL"
+				v-tooltip.bottom="t(livePreviewMode === null ? 'live_preview.enable' : 'live_preview.disable')"
+				rounded
+				icon
+				class="action-preview"
+				:secondary="livePreviewMode === null"
+				@click="toggleSplitView"
+			>
+				<v-icon :name="livePreviewMode === null ? 'visibility' : 'visibility_off'" outline />
+			</v-button>
```
I think this button should be hidden for tablet and smaller screen sizes. While you can trigger the split screen on smaller screens currently this does not work. 
![image](https://github.com/directus/directus/assets/9389634/9b77c5e4-d902-45b7-aaa1-c3c7fa7238a2)

br41nslug commented on app/src/modules/content/components/live-preview.vue:
```diff
@@ -0,0 +1,247 @@
+	isRefreshing.value = false;
+}
+
+(window as any).refreshLivePreview = refresh;
```
Instead of typecasting custom properties can be added to the window like so in TS:
```suggestion
declare global {
	interface Window {
		refreshLivePreview: () => void;
	}
}

window.refreshLivePreview = refresh;
```

azrikahar replied:
Good catch! Seems like there's actually the same approach for another custom property in:

https://github.com/directus/directus/blob/4f768ca346fd4ce12c08884bf90de09b22cb4bdb/packages/composables/src/use-element-size.ts#L5-L9

implemented the suggested change in https://github.com/directus/directus/pull/17794/commits/4f768ca346fd4ce12c08884bf90de09b22cb4bdb.

br41nslug replied:
Awesome don't forget to clean up the remaining `(window as any)` references in the source 🚀 😄

paescuj committed:
Move debounce to private-view

phazonoverload committed:
Update Nuxt 3 guides - styling and sentence clarity.

azrikahar committed:
add refreshLivePreview custom property to window

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

estheragbaje commented on docs/guides/headless-cms/live-preview/nuxt-3.md:
```diff
@@ -0,0 +1,153 @@
+
+Through this guide, you have successfully set up the live preview feature in Directus for your Nuxt 3 project.
+
+Share the preview URL with your team members, stakeholders, or clients to allow them to see how content changes would look.
```
Solid docs🎉 Everything reads good.

azrikahar committed:
remove remaining "window as any" typecasts

paescuj committed:
Show 'grab-bar' only if wrapper is in viewport

azrikahar committed:
toggle preview instead of split in mobile view

azrikahar committed:
tweak to unwrap nested IF conditions

LasseRosenow commented:
![image](https://github.com/directus/directus/assets/10547444/3b7e13a4-077f-40cb-ad6e-bb9bf7b34762)

Actually I must say I find this colored but also crossed out eye really confusing.
Because usually it is not colored, so that means colored = enabled
But then the icon changes to crossed out ... so .. crossed out eye is enabled?

I think those 2 solutions for this problem cannot be mixed.
Either use color to show the state, like is the Eye enabled or not?
Or switch the icon to show its state.

But both together is confusing I think.

rijkvanzanten commented:
Fully agree @LasseRosenow 👍🏻 The icon shouldn't change when it's activated 🙂

azrikahar committed:
do not change preview button icon when enabled

azrikahar commented:
>  The icon shouldn't change when it's activated

Fixed in https://github.com/directus/directus/pull/17794/commits/755a4a27ac423bb113f6c223c5d79093ede9a2ea 👍

azrikahar committed:
hide grab-bar in mobile

azrikahar committed:
check main content width instead of window width

licitdev committed:
Fix save-and-stay not updating preview url

licitdev committed:
Update refresh button function call

paescuj committed:
Add frame loader to prevent flashing

paescuj committed:
Prevent scrollbar flashing when toggling splitview

azrikahar committed:
lift specific logic from v-resizable to private-view

This piece of logic was meant only for the content area and shouldn't affect every v-resizable usages

azrikahar committed:
use v-show instead of v-if to hide main content

v-resizable has an internalWidth that should not be reset when user resize the left navigation bar, as well as the form shouldn't be re-rendered

azrikahar committed:
check target to hide the grab-bar in toggle mode

as we're only hiding the target rather than disabling the wrapper, we had to check whether the target is visible as well to determine whether to show the grab-bar

paescuj committed:
Fix zoom behaviour & remove leftovers

azrikahar committed:
remove wrapperIsVisible check

in favor of only relying on targetIsVisible check

paescuj committed:
Revert "Add frame loader to prevent flashing"

This reverts commit e4c2d3d4bdb59c1d955f5a742867dda94657fa59.

paescuj commented on api/src/database/migrations/20230313A-add-preview-settings.ts:
Yes, let's do that 👍

paescuj committed:
Final tweaks

- Enhance frame zoom
- Min width for frame
- Hide main content based on overall size
  (ensure it works for all browser sizes)
- Various enhancements & fixes

paescuj committed:
Merge branch 'main' into live-preview

paescuj committed:
Update migration date

br41nslug committed:
Merge branch 'main' into live-preview

paescuj committed:
Create rare-bananas-melt.md

paescuj commented on app/src/views/private/private-view.vue:
```diff
@@ -145,34 +155,78 @@ const contentEl = ref<HTMLElement>();
+		return true;
+	}
+
+	let remainigPlace;
```
Yeah, makes more sense 😄 Will update 👍

paescuj committed:
Better naming for remaining width var

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>
